Optionally send min and max metrics for OTLP Histograms when available

mx-psi opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

- Bump opentelemetry-mapping-go modules to v0.1.4.
- Deprecate `otlp_config.metrics.histograms.send_count_sum_metrics` in favor of `otlp_config.metrics.histograms.send_aggregation_metrics`.


### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

Most OpenTelemetry SDKs today support sending min and max values for OTLP Histograms and OTLP Exponential Histograms.
When in delta aggregation temporality, we can produce min and max metrics that can be visualized in the backend.

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

Counterpart change in OpenTelemetry Collector: open-telemetry/opentelemetry-collector-contrib#20285

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

Configure OTLP ingest to use the new flag:

```
otlp_config:
  receiver:
    protocols:
      grpc:
  metrics:
    histograms:
      send_aggregation_metrics: true
```

Send a [delta](https://docs.datadoghq.com/opentelemetry/guide/otlp_delta_temporality/) OTLP Histogram to the OTLP ingest endpoint and check that `.min` and `.max` metrics are sent to Datadog.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

mx-psi committed:
Bump opentelemetry-mapping-go modules to v0.1.4

mx-psi committed:
Deprecate `otlp_config.metrics.histograms.send_count_sum_metrics` in favor of `otlp_config.metrics.histograms.send_aggregation_metrics`

ajgajg1134 reviewed (approved):
Looks good from Agent APM!

hush-hush commented on pkg/config/config_template.yaml:
```diff
@@ -3469,12 +3469,19 @@ api_key:
       # mode: distributions
 
+      ## Deprecated - use `send_aggregation_metrics` instead
       ## @param send_count_sum_metrics - boolean - optional - default: false
       ## @env DD_OTLP_CONFIG_METRICS_HISTOGRAMS_SEND_COUNT_SUM_METRICS - boolean - optional - default: false
-      ## Whether to report sum and count as separate histogram metrics.
+      ## Whether to report sum, count, min and max as separate histogram metrics.
       #
       # send_count_sum_metrics: false
 
+      ## @param send_aggregation_metrics - boolean - optional - default: false
+      ## @env DD_OTLP_CONFIG_METRICS_HISTOGRAMS_SEND_AGGREGATION_METRICS - boolean - optional - default: false
+      ## Whether to report sum, count, min and max as separate histogram metrics.
+      #
+      # send_aggregation_metrics: false
+
```
I think we should explain which on will take precedence when both are set.

mx-psi replied:
> I think so, having to deploy new configuration and restart agent to detect errors is not a great workflow for customers.

I am used to the OpenTelemetry Collector way of doing things, where the approach is 'fail fast if anything is wrong/ambiguous'. I think it works great there and this approach helps us prevent hard-to-detect faulty deployments, but I don't want to be inconsistent with how the rest of the Agent works just because I am used to that.

> In the past, for similar cases, we gave precedence on the old value and printed a warning.

I will do that then :+1:

mx-psi replied:
Fixed in e5a6b34e8d418048a634cbeb8d2988f45a5c5940

rtrieu commented on pkg/config/config_template.yaml:
```diff
@@ -3469,12 +3469,19 @@ api_key:
       ## @param send_count_sum_metrics - boolean - optional - default: false
       ## @env DD_OTLP_CONFIG_METRICS_HISTOGRAMS_SEND_COUNT_SUM_METRICS - boolean - optional - default: false
-      ## Whether to report sum and count as separate histogram metrics.
+      ## Whether to report sum, count, min and max as separate histogram metrics.
```
```suggestion
      ## Whether to report sum, count, min, and max as separate histogram metrics.
```

rtrieu commented on pkg/config/config_template.yaml:
```diff
@@ -3469,12 +3469,19 @@ api_key:
 
+      ## @param send_aggregation_metrics - boolean - optional - default: false
+      ## @env DD_OTLP_CONFIG_METRICS_HISTOGRAMS_SEND_AGGREGATION_METRICS - boolean - optional - default: false
+      ## Whether to report sum, count, min and max as separate histogram metrics.
```
```suggestion
      ## Whether to report sum, count, min, and max as separate histogram metrics.
```

rtrieu reviewed (approved):
Minor suggestion, but otherwise looks good!

mx-psi committed:
Fail if both send_count_sum_metrics and send_aggregation_metrics are set at the same time

mx-psi committed:
Apply suggestions from code review

Co-authored-by: Rosa Trieu <107086888+rtrieu@users.noreply.github.com>

mx-psi committed:
Address PR comments

We now override send_aggregation_metrics with the value of send_count_sum_metrics if both are set.
To do this we leverage the custom unmarshaling system.

mx-psi commented:
@hush-hush ping, please take another look!
